rust/src/mysql/mysql.rs

```diff
@@ -0,0 +1,1133 @@
+            affected_rows: None,
+            tls: None,
+            tx_data: AppLayerTxData::new(),
+            complete: false,
+            rows: None,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum MysqlStateProgress {
+    // Default State
+    Init,
+
+    // Connection Phase
+    // Server send HandshakeRequest to Client
+    Handshake,
+    // Client send HandshakeResponse to Server
+    // Server send AuthSwitchRequest to Client
+    Auth,
+    // Server send OkResponse to Client
+    AuthFinished,
+
+    // Command Phase
+    // Client send QueryRequest to Server
+    CommandReceived,
+    // Server send EOF with 0x0A to Client
+    TextResulsetContinue,
+    // Server send QueryResponse to Client or Ok to Client
+    CommandResponseReceived,
+    // Server send LocalFileRequest with zero length to Client
+    LocalFileRequestReceived,
+    // Client send empty packet to Server
+    LocalFileContentFinished,
+    // Client send StatisticsRequest to Server
+    StatisticsReceived,
+    // Server send StatisticsResponse to client
+    StatisticsResponseReceived,
+    // Client send FieldList to Server
+    FieldListReceived,
+    // Server send FieldListResponse to client
+    FieldListResponseReceived,
+    // Client send ChangeUserRequest to Server
+    ChangeUserReceived,
+    // Server send OkResponse to client
+    ChangeUserResponseReceived,
+    // Client send Unknown to Server
+    UnknownCommandReceived,
+    // Client send StmtPrepareRequest to Server
+    StmtPrepareReceived,
+    // Server send StmtPrepareResponse to Client
+    StmtPrepareResponseReceived,
+    // Client send StmtExecRequest to Server
+    StmtExecReceived,
+    // Server send StmtExecResponse with EOF status equal 0x0a to Client
+    StmtExecResponseContinue,
+    // Server send ResultSetResponse to Client or Ok Response to Client
+    StmtExecResponseReceived,
+    // Client send StmtFetchRequest to Server
+    StmtFetchReceived,
+    // Server send StmtFetchResponse with EOF status equal 0x0a to Client
+    StmtFetchResponseContinue,
+    // Server send StmtFetchResponse to Client
+    StmtFetchResponseReceived,
+    // Client send StmtFetchRequest to Server
+    StmtResetReceived,
+    // Server send Ok or Err Response to Client
+    StmtResetResponseReceived,
+    // Client send StmtCloseRequest to Server
+    StmtCloseReceived,
+
+    // Client send QueryRequest and command equal Quit to Server
+    // Client send ChangeUserRequest to server and server send ErrResponse to Client
+    // Transport Layer EOF
+    // Transport Layer Upgrade to TLS
+    Finished,
+}
+
+#[derive(Debug)]
+pub struct MysqlState {
+    pub state_data: AppLayerStateData,
+    pub tx_id: u64,
+    transactions: VecDeque<MysqlTransaction>,
+    request_gap: bool,
+    response_gap: bool,
+    state_progress: MysqlStateProgress,
+    tx_index_completed: usize,
+
+    client_flags: u32,
+    version: Option<String>,
+    tls: bool,
```

Comment:
Do you handle protocol change ? (like `STARTTLS` does in SMTP)

Reply:
Yes, Please check this.
 https://github.com/Kotodian/suricata/blob/becd84350a53a7b16de80c135d5263908b7b6db1/rust/src/mysql/mysql.rs#L326-L330